### PR TITLE
Draft Abstract Staking logic

### PR DESCRIPTION
**Pull request by mllwchrry**
- [ ] Since this PR suggests a **bug fix**, the tests have been added and the coverage is 100%.
- [x] Since this PR introduces a **new feature**, the update has been discussed in an Issue or with the team.
- [ ] This PR is just a **minor change**, like a typo fix.

---

### CODE REVIEW

File: contracts/staking/ValueDistributor.sol [hunk 1]
```diff
@@ -3,21 +3,81 @@ pragma solidity ^0.8.4;
...
 
-    mapping(address => UserDistribution) public userDistributions;
+    mapping(address => UserDistribution) private _userDistributions;
 
+    /**
+     * @notice Distributes value to a specific user.
+     * @param user_ The address of the user.
+     * @param amount_ The amount of value to distribute.
+     */
+    function distributeValue(address user_, uint256 amount_) public {
```
Comment (Arvolear) [inline contracts/staking/ValueDistributor.sol:35]: Should this call `_update`?

File: contracts/staking/ValueDistributor.sol [hunk 2]
```diff
@@ -0,0 +1,208 @@
...
+
+    mapping(address => UserDistribution) private _userDistributions;
+
+    /**
+     * @notice Distributes value to a specific user.
+     * @param user_ The address of the user.
+     * @param amount_ The amount of value to distribute.
+     */
+    function distributeValue(address user_, uint256 amount_) public {
+        require(amount_ > 0, "Amount has to be more than 0");
```
Comment (Arvolear) [inline contracts/staking/ValueDistributor.sol:36]: Let's name errors with "<contract name>: " prefix

File: contracts/staking/ValueDistributor.sol [hunk 3]
```diff
@@ -0,0 +1,208 @@
...
+    uint256 private _updatedAt;
+
+    mapping(address => UserDistribution) private _userDistributions;
+
+    /**
+     * @notice Distributes value to a specific user.
+     * @param user_ The address of the user.
+     * @param amount_ The amount of value to distribute.
+     */
+    function distributeValue(address user_, uint256 amount_) public {
```
Comment (Arvolear) [inline contracts/staking/ValueDistributor.sol:35]: This has to be `internal`

File: contracts/staking/Staking.sol [hunk 4]
```diff
@@ -0,0 +1,186 @@
...
+    }
+
+    /**
+     * @notice Initializes the contract setting the values provided by the deployer as shares token, rewards token, reward rate and staking start time.
+     * @param sharesToken_ The address of the shares token.
+     * @param rewardsToken_ The address of the rewards token.
+     * @param rate_ The reward rate.
+     * @param stakingStartTime_ The staking start time
+     */
+    constructor(
```
Comment (Arvolear) [inline contracts/staking/Staking.sol:42]: Let's make this an init function (check the other contracts in the repo)

File: contracts/staking/Staking.sol [hunk 5]
```diff
@@ -0,0 +1,186 @@
...
+     *
+     * Note: All the rewards are claimed after the shares were removed.
+     *
+     * @param user_ The address of the user.
+     * @param amount_ The amount of shares removed.
+     */
+    function _afterRemoveShares(address user_, uint256 amount_) internal virtual override {
+        SafeERC20.safeTransfer(IERC20(_sharesToken), user_, amount_);
+
+        distributeValue(msg.sender, getOwedValue(user_));
```
Comment (Arvolear) [inline contracts/staking/Staking.sol:155]: Do we need to call this here?

File: contracts/staking/Staking.sol [hunk 6]
```diff
@@ -0,0 +1,186 @@
...
+     */
+    function stake(uint256 amount_) public stakingStarted {
+        _addShares(msg.sender, amount_);
+    }
+
+    /**
+     * @notice Withdraws the specified amount of tokens.
+     * @param amount_ The amount of tokens to withdraw.
+     */
+    function withdraw(uint256 amount_) public stakingStarted {
```
Comment (Arvolear) [inline contracts/staking/Staking.sol:70]: Let's rename this to `unstake`

Comment (Arvolear) [reply contracts/staking/Staking.sol:70]: And add `withdraw` function that would call `unstake` and `claim` under the hood

File: contracts/staking/ValueDistributor.sol [hunk 7]
```diff
@@ -119,6 +108,25 @@ abstract contract ValueDistributor {
...
+     * @notice Distributes value to a specific user.
+     * @param user_ The address of the user.
+     * @param amount_ The amount of value to distribute.
+     */
+    function _distributeValue(address user_, uint256 amount_) internal {
+        require(amount_ > 0, "ValueDistributor: amount has to be more than 0");
+        require(
+            amount_ <= _userDistributions[user_].owedValue,
+            "ValueDistributor: insufficient amount"
+        );
```
Comment (Arvolear) [inline contracts/staking/ValueDistributor.sol:121]: This require won't work because the owedValue is updated on the next line

Comment (Arvolear) [reply contracts/staking/ValueDistributor.sol:121]: Just call _update before the requires

File: contracts/staking/Staking.sol [hunk 8]
```diff
@@ -0,0 +1,193 @@
...
+    function unstake(uint256 amount_) public stakingStarted {
+        _removeShares(msg.sender, amount_);
+    }
+
+    /**
+     * @notice Claims the specified amount of rewards.
+     * @param amount_ The amount of rewards to claim.
+     */
+    function claim(uint256 amount_) public stakingStarted {
+        uint256 owed = userDistribution(msg.sender).owedValue;
```
Comment (Arvolear) [inline contracts/staking/Staking.sol:91]: Same issue here

File: contracts/staking/Staking.sol [hunk 9]
```diff
@@ -0,0 +1,193 @@
...
+        SafeERC20.safeTransfer(IERC20(_sharesToken), user_, amount_);
+    }
+
+    /**
+     * @notice Hook function that is called after value has been distributed to a user.
+     * @param user_ The address of the user.
+     * @param amount_ The amount of value distributed.
+     */
+    function _afterDistributeValue(address user_, uint256 amount_) internal virtual override {
+        SafeERC20.safeTransfer(IERC20(_rewardsToken), user_, amount_);
```
Comment (Arvolear) [inline contracts/staking/Staking.sol:171]: Let's make the code more attractive with `using` directives. Then you will be able to call IERC20(_rewardsToken).safeTransfer directly

File: contracts/staking/Staking.sol [hunk 10]
```diff
@@ -0,0 +1,193 @@
...
+     */
+    function __Staking_init(
+        address sharesToken_,
+        address rewardsToken_,
+        uint256 rate_,
+        uint256 stakingStartTime_
+    ) internal onlyInitializing {
+        require(sharesToken_ != address(0), "Staking: zero address cannot be the Shares Token");
+        require(rewardsToken_ != address(0), "Staking: zero address cannot be the Rewards Token");
+        require(rate_ > 0, "Staking: rate has to be more than 0");
```
Comment (Arvolear) [inline contracts/staking/Staking.sol:51]: Let's allow rate to be 0

File: contracts/staking/Staking.sol [hunk 11]
```diff
@@ -0,0 +1,193 @@
...
+    uint256 private _rate;
+
+    uint256 private _stakingStartTime;
+
+    /**
+     * @dev Throws if the staking has not started yet.
+     */
+    modifier stakingStarted() {
+        _checkStakingStarted();
+
```
Comment (Arvolear) [inline contracts/staking/Staking.sol:32]:
```suggestion
```

File: contracts/staking/ValueDistributor.sol [hunk 12]
```diff
@@ -0,0 +1,216 @@
...
+
+        _afterRemoveShares(user_, amount_);
+    }
+
+    /**
+     * @notice Distributes value to a specific user.
+     * @param user_ The address of the user.
+     * @param amount_ The amount of value to distribute.
+     */
+    function _distributeValue(address user_, uint256 amount_) internal {
```
Comment (Arvolear) [inline contracts/staking/ValueDistributor.sol:116]: virtual?

File: contracts/staking/AbstractValueDistributor.sol [hunk 13]
```diff
@@ -0,0 +1,211 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.4;
+
+import {PRECISION, DECIMAL} from "../utils/Globals.sol";
+
+/**
+ * @notice The AbstractValueDistributor module
+ *
+ * Contract module for distributing a value among users based on their shares.
```
Comment (Arvolear) [inline contracts/staking/AbstractValueDistributor.sol:9]:
```suggestion
 * Contract module for distributing value among users based on their shares.
```

File: contracts/staking/AbstractValueDistributor.sol [hunk 14]
```diff
@@ -0,0 +1,211 @@
...
+ *
+ * This contract can be used as a base contract for implementing various distribution mechanisms,
+ * such as token staking, profit sharing, or dividend distribution.
+ *
```
Comment (Arvolear) [inline contracts/staking/AbstractValueDistributor.sol:13]: Please add a brief explanation of the algorithm

File: contracts/staking/AbstractValueDistributor.sol [hunk 15]
```diff
@@ -0,0 +1,211 @@
...
+            userDist.owedValue;
+    }
+
+    /**
+     * @notice Adds shares to a user's distribution.
+     * @param user_ The address of the user.
+     * @param amount_ The amount of shares to add.
+     */
+    function _addShares(address user_, uint256 amount_) internal virtual {
+        require(amount_ > 0, "ValueDistributor: amount has to be more than 0");
```
Comment (Arvolear) [inline contracts/staking/AbstractValueDistributor.sol:81]: Let's check that the user is not address(0)

File: contracts/staking/AbstractValueDistributor.sol [hunk 16]
```diff
@@ -0,0 +1,211 @@
...
+
+        _afterRemoveShares(user_, amount_);
+    }
+
+    /**
+     * @notice Distributes value to a specific user.
+     * @param user_ The address of the user.
+     * @param amount_ The amount of value to distribute.
+     */
+    function _distributeValue(address user_, uint256 amount_) internal virtual {
```
Comment (Arvolear) [inline contracts/staking/AbstractValueDistributor.sol:116]: Same user check here

File: contracts/staking/AbstractValueDistributor.sol [hunk 17]
```diff
@@ -0,0 +1,211 @@
...
+
+            userDist.owedValue +=
+                (userDist.shares * (_cumulativeSum - userDist.cumulativeSum)) /
+                PRECISION;
+            userDist.cumulativeSum = _cumulativeSum;
+        }
+    }
+
+    /**
+     * @notice Gets the value to be distributed for a given time period.
```
Comment (Arvolear) [inline contracts/staking/AbstractValueDistributor.sol:185]: Please add a note that users will usually be required to override this function to provide custom distribution mechanics

File: contracts/staking/AbstractValueDistributor.sol [hunk 18]
```diff
@@ -0,0 +1,211 @@
...
+        uint256 timeLastUpdate
+    ) internal view virtual returns (uint256) {
+        return DECIMAL * (timeUpTo - timeLastUpdate); // 1 token with 18 decimals per second
+    }
+
+    /**
+     * @notice Gets the expected cumulative sum of values that have been distributed at a given timestamp.
+     * @param timeUpTo The end timestamp.
+     * @return The future cumulative sum of tokens that have been distributed.
+     */
```
Comment (Arvolear) [inline contracts/staking/AbstractValueDistributor.sol:201]:
```suggestion
    /**
     * @notice Gets the expected cumulative sum of value per token staked distributed at a given timestamp.
     * @param timeUpTo The timestamp up to which to calculate the value distribution 
     * @return The future cumulative sum of value per token staked that has been distributed.
     */
```

File: contracts/staking/AbstractStaking.sol [hunk 19]
```diff
@@ -0,0 +1,194 @@
...
+     * @param amount_ The amount of tokens to stake.
+     */
+    function stake(uint256 amount_) public stakingStarted {
+        _addShares(msg.sender, amount_);
+    }
+
+    /**
+     * @notice Withdraws all the staked tokens.
+     *
+     * Note: All the rewards are claimed after the shares were removed.
```
Comment (Arvolear) [inline contracts/staking/AbstractStaking.sol:75]:
```suggestion
     * Note: All the rewards are claimed after the shares are removed.
```

File: contracts/staking/AbstractStaking.sol [hunk 20]
```diff
@@ -0,0 +1,194 @@
...
+    /**
+     * @notice Stakes the specified amount of tokens.
+     * @param amount_ The amount of tokens to stake.
+     */
+    function stake(uint256 amount_) public stakingStarted {
+        _addShares(msg.sender, amount_);
+    }
+
+    /**
+     * @notice Withdraws all the staked tokens.
```
Comment (Arvolear) [inline contracts/staking/AbstractStaking.sol:73]:
```suggestion
     * @notice Withdraws all the staked tokens together with rewards
```

File: contracts/staking/AbstractStaking.sol [hunk 21]
```diff
@@ -0,0 +1,194 @@
...
+     *
+     * Note: All the rewards are claimed after the shares were removed.
+     */
+    function withdraw() public stakingStarted {
```
Comment (Arvolear) [inline contracts/staking/AbstractStaking.sol:77]: Can you please place this function after the `claim` function?

Comment (Arvolear) [reply contracts/staking/AbstractStaking.sol:77]: Let's return `userDistribution(msg.sender).shares` and  `getOwedValue(msg.sender)` from this function

File: contracts/staking/AbstractStaking.sol [hunk 22]
```diff
@@ -0,0 +1,194 @@
...
+     * @notice Returns the rewards token.
+     * @return The address of the rewards token contract.
+     */
+    function rewardsToken() public view returns (address) {
+        return _rewardsToken;
+    }
+
+    /**
+     * @notice Returns the staking start time.
+     * @return The timestamp when staking started.
```
Comment (Arvolear) [inline contracts/staking/AbstractStaking.sol:116]:
```suggestion
     * @return The timestamp when staking starts.
```

File: test/staking/AbstractStaking.test.ts [hunk 23]
```diff
@@ -0,0 +1,434 @@
...
+    it("should handle the whole staling process using 6-decimal values", async () => {
+      const AbstractStakingMock = await ethers.getContractFactory("AbstractStakingMock");
+      const ERC20Mock = await ethers.getContractFactory("ERC20Mock");
+
+      const abstractStaking6Decimals = await AbstractStakingMock.deploy();
+      const SHARES_TOKEN = await ERC20Mock.deploy("SharesMock", "SMock", 6);
+      const REWARDS_TOKEN = await ERC20Mock.deploy("RewardsMock", "RMock", 6);
+      await REWARDS_TOKEN.mint(await abstractStaking6Decimals.getAddress(), wei(1000));
+
+      await abstractStaking6Decimals.__AbstractStakingMock_init(SHARES_TOKEN, REWARDS_TOKEN, wei(1, 6), 3n);
```
Comment (Arvolear) [inline test/staking/AbstractStaking.test.ts:382]: You can use an existing abstract staking contract for that. Then call performStakingManipulations

File: contracts/mock/staking/AbstractStakingMock.sol [hunk 24]
```diff
@@ -0,0 +1,76 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.4;
+
+import "@openzeppelin/contracts/utils/Multicall.sol";
```
Comment (Arvolear) [inline contracts/mock/staking/AbstractStakingMock.sol:4]: Named imports

File: contracts/mock/staking/AbstractStakingMock.sol [hunk 25]
```diff
@@ -0,0 +1,76 @@
...
+        return userDistribution(user_).owedValue;
+    }
+}
+
+contract StakersFactory is Multicall {
+    Staker[] public stakers;
+
+    function createStaker() public {
+        Staker staker = new Staker();
+        stakers.push(staker);
```
Comment (Arvolear) [inline contracts/mock/staking/AbstractStakingMock.sol:50]:
```suggestion
        Staker staker_ = new Staker();
        stakers.push(staker_);
```

File: contracts/mock/staking/AbstractStakingMock.sol [hunk 26]
```diff
@@ -0,0 +1,76 @@
...
+        Staker(staker_)._stake(stakingContract_, token_, amount_);
+    }
+
+    function unstake(address stakingContract_, address staker_, uint256 amount_) external {
+        Staker(staker_)._unstake(stakingContract_, amount_);
+    }
+}
+
+contract Staker {
+    function _stake(address stakingContract_, address token_, uint256 amount_) public {
```
Comment (Arvolear) [inline contracts/mock/staking/AbstractStakingMock.sol:68]:
```suggestion
    function stake(address stakingContract_, address token_, uint256 amount_) public {
```

File: contracts/mock/staking/AbstractStakingMock.sol [hunk 27]
```diff
@@ -0,0 +1,76 @@
...
+        IERC20(token_).approve(stakingContract_, amount_);
+        AbstractStakingMock(stakingContract_).stake(amount_);
+    }
+
+    function _unstake(address stakingContract_, uint256 amount_) public {
```
Comment (Arvolear) [inline contracts/mock/staking/AbstractStakingMock.sol:73]:
```suggestion
    function unstake(address stakingContract_, uint256 amount_) external {
```

File: contracts/mock/staking/AbstractValueDistributorMock.sol [hunk 28]
```diff
@@ -0,0 +1,45 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.4;
+
+import "@openzeppelin/contracts/utils/Multicall.sol";
```
Comment (Arvolear) [inline contracts/mock/staking/AbstractValueDistributorMock.sol:4]: Named import

File: contracts/staking/AbstractStaking.sol [hunk 29]
```diff
@@ -0,0 +1,200 @@
...
+     *
+     * @return shares_ The amount of shares being withdrawn.
+     * @return owedValue_ The total value of the rewards owed to the user.
+     */
+    function withdraw() public stakingStarted returns (uint256 shares_, uint256 owedValue_) {
+        shares_ = userDistribution(msg.sender).shares;
+        owedValue_ = getOwedValue(msg.sender);
+
+        unstake(userDistribution(msg.sender).shares);
+        claim(getOwedValue(msg.sender));
```
Comment (Arvolear) [inline contracts/staking/AbstractStaking.sol:101]:
```suggestion
        unstake(shares_);
        claim(owedValue_);
```

File: test/staking/AbstractStaking.test.ts [hunk 30]
```diff
@@ -0,0 +1,654 @@
...
+  const reverter = new Reverter();
+
+  let FIRST: SignerWithAddress;
+  let SECOND: SignerWithAddress;
+  let THIRD: SignerWithAddress;
+
+  let SHARES_TOKEN: ERC20Mock;
+  let REWARDS_TOKEN: ERC20Mock;
+
+  let shares_decimals: number;
```
Comment (Arvolear) [inline test/staking/AbstractStaking.test.ts:20]: camelCase please

File: test/staking/AbstractStaking.test.ts [hunk 31]
```diff
@@ -0,0 +1,654 @@
...
+  let rate: bigint;
+
+  let abstractStaking: AbstractStakingMock;
+
+  const mintAndApproveTokens = async (user: SignerWithAddress, token: ERC20Mock, amount: bigint) => {
+    await token.mint(user, amount);
+    await token.connect(user).approve(abstractStaking, amount);
+  };
+
+  const performStakingManipulations = async () => {
```
Comment (Arvolear) [inline test/staking/AbstractStaking.test.ts:33]: This function is only used once. Please move its contents directly to the test

File: contracts/staking/AbstractValueDistributor.sol [hunk 32]
```diff
@@ -0,0 +1,231 @@
...
+     *
+     * @param timeUpTo The end timestamp of the period.
+     * @param timeLastUpdate The start timestamp of the period.
+     * @return The value to be distributed for the period.
+     */
+    function _getValueToDistribute(
+        uint256 timeUpTo,
+        uint256 timeLastUpdate
+    ) internal view virtual returns (uint256) {
+        return DECIMAL * (timeUpTo - timeLastUpdate); // 1 token with 18 decimals per second
```
[the remark this answers is not in the document]
Comment (Arvolear) [reply contracts/staking/AbstractValueDistributor.sol:214]: Yeah, let's make it purely virtual

File: contracts/staking/AbstractStaking.sol [hunk 33]
```diff
@@ -0,0 +1,200 @@
...
+     * @param stakingStartTime_ The staking start time
+     */
+    function __AbstractStaking_init(
+        address sharesToken_,
+        address rewardsToken_,
+        uint256 rate_,
+        uint256 stakingStartTime_
+    ) internal onlyInitializing {
+        require(sharesToken_ != address(0), "Staking: zero address cannot be the Shares Token");
+        require(rewardsToken_ != address(0), "Staking: zero address cannot be the Rewards Token");
```
Comment (dovgopoly) [inline contracts/staking/AbstractStaking.sol:56]: When shares and reward tokens are equal, we may accidently withdraw other users shares as a reward. Please, leave warning if it is an expected behavior.

File: contracts/staking/AbstractValueDistributor.sol [hunk 34]
```diff
@@ -0,0 +1,231 @@
...
+    }
+
+    /**
+     * @notice Gets the expected cumulative sum of value per token staked distributed at a given timestamp.
+     * @param timeUpTo The timestamp up to which to calculate the value distribution.
+     * @return The future cumulative sum of value per token staked that has been distributed.
+     */
+    function _getFutureCumulativeSum(uint256 timeUpTo) internal view returns (uint256) {
```
[the remark this answers is not in the document]
Comment (Arvolear) [reply contracts/staking/AbstractValueDistributor.sol:222]: Like, adding the require here still won't save the user from shooting themselves in the leg

[the remark this answers is not in the document]
Comment (dovgopoly) [reply contracts/staking/AbstractValueDistributor.sol:222]: They will get something more comprehensive than an overflow error though.

[the remark this answers is not in the document]
Comment (dovgopoly) [reply contracts/staking/AbstractValueDistributor.sol:222]: Maybe we don't need the timeUpTo_ parameter at all, if we don't consider scenarios where this function is not called with block.timestamp.

File: contracts/staking/AbstractStaking.sol (hunk 33 above)
Comment (Arvolear) [reply contracts/staking/AbstractStaking.sol:56]: Let's leave a comment on that. Virtually it is not a smart contract issue

File: contracts/staking/AbstractValueDistributor.sol (hunk 34 above)
[the remark this answers is not in the document]
Comment (Arvolear) [reply contracts/staking/AbstractValueDistributor.sol:222]: But the overflow will only happen inside `_getValueToDistribute()` function which is purely virtual

[the remark this answers is not in the document]
Comment (Arvolear) [reply contracts/staking/AbstractValueDistributor.sol:222]: Meaning that the user has to modify the flow somehow to force the underflow

File: contracts/staking/AbstractStaking.sol [hunk 35]
```diff
@@ -41,6 +41,9 @@ abstract contract AbstractStaking is AbstractValueDistributor, Initializable {
 
     /**
      * @notice Initializes the contract setting the values provided as shares token, rewards token, reward rate and staking start time.
+     *
+     * Warning: when shares and rewards tokens are the same, users may accidentally withdraw other users' shares as a reward.
```
Comment (Arvolear) [inline contracts/staking/AbstractStaking.sol:45]: ... if the reward token balance is improperly handled.

File: contracts/staking/AbstractValueDistributor.sol [hunk 36]
```diff
@@ -0,0 +1,229 @@
...
+     */
+    function getOwedValue(address user_) public view returns (uint256) {
+        UserDistribution storage userDist = _userDistributions[user_];
+
+        return
+            (userDist.shares *
+                (_getFutureCumulativeSum(block.timestamp) - userDist.cumulativeSum)) /
+            PRECISION +
+            userDist.owedValue;
+    }
```
Comment (Hrom131) [inline contracts/staking/AbstractValueDistributor.sol:83]: Maybe we should add another function that will accept the timestamp for which we want to calculate the owed value

**Review by Arvolear — APPROVED** [review]
LGTM!

**Pull request merged**